### PR TITLE
[bugfix]Remember the c++ plugins folders in Options dialog

### PR DESCRIPTION
**Pull request by DelazJ**
and enable the remove button
(Currently folders added to the list of c++ plugins are cleared when closing the dialog)

### CODE REVIEW

File: src/app/qgsoptions.cpp [hunk 1]
```diff
@@ -255,14 +255,18 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl, const QList<QgsOpti
...
   //local directories to search when loading c++ plugins
   QStringList pathList = mSettings->value( QStringLiteral( "plugins/searchPathsForPlugins" ) ).toStringList();
-  Q_FOREACH ( const QString &path, pathList )
+  if ( !pathList.isEmpty() )
   {
-    QListWidgetItem *newItem = new QListWidgetItem( mListPluginPaths );
-    newItem->setText( path );
-    newItem->setFlags( Qt::ItemIsEditable | Qt::ItemIsEnabled | Qt::ItemIsSelectable );
-    mListPluginPaths->addItem( newItem );
+    Q_FOREACH ( const QString &path, pathList )
```
Comment (m-kuhn) [inline src/app/qgsoptions.cpp]: @DelazJ since you are touching this code

```
const QStringList pathList...

for ( const QString &path : pathList )
```

Comment (DelazJ) [reply src/app/qgsoptions.cpp]: @m-kuhn the `if ( !pathList.isEmpty() )` check is then unnecessary? is it implicite?
That said, giving a look to the different list/tree widgets in the file, it's not always used...

Comment (m-kuhn) [reply src/app/qgsoptions.cpp]: Yes, if there is nothing in the pathList, the loop will be executed 0 times, implicitly (that's the case with `Q_FOREACH` as well as `for ( : )`

Comment (DelazJ) [reply src/app/qgsoptions.cpp]: pathList is used for different lists in the file, hence setting it as const makes the build fail

Comment (m-kuhn) [reply src/app/qgsoptions.cpp]: I see. If you want to make it nice, rename those other variables. If you want this to get merged with as little effort as possible, remove the `const` and use it in each loop as `for ( const QString &path : qgis::as_const( pathList ) )`.

File: src/app/qgsoptions.cpp [hunk 2]
```diff
@@ -782,11 +782,11 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl, const QList<QgsOpti
...
 
   // predefined scales for scale combobox
-  QString myPaths = mSettings->value( QStringLiteral( "Map/scales" ), PROJECT_SCALES ).toString();
-  if ( !myPaths.isEmpty() )
+  QString scalePaths = mSettings->value( QStringLiteral( "Map/scales" ), PROJECT_SCALES ).toString();
+  if ( !scalePaths.isEmpty() )
   {
-    QStringList myScalesList = myPaths.split( ',' );
-    Q_FOREACH ( const QString &scale, myScalesList )
+    QStringList ScalesList = scalePaths.split( ',' );
```
Comment (m-kuhn) [inline src/app/qgsoptions.cpp]: watch out, variable names first character is lower-case (class names are capital first letter)

Comment (m-kuhn) [reply src/app/qgsoptions.cpp]: and also watch out: if using this `for ( : )` construct, the container part needs to be `const`.

Comment (DelazJ) [reply src/app/qgsoptions.cpp]: > if using this for ( : ) construct, the container part needs to be const.

That means that i should rename the other occurrences of pathList, ie for svg, composer templates, hidden path...?

Comment (m-kuhn) [reply src/app/qgsoptions.cpp]: Easiest: `const QStringList scalesList = ...`, with the implication that different names will need to be used for different variables (which I think makes sense, they *are* different lists after all).
Or: `for ( const T &var : qgis::as_const( container ) )`

caveat: there are also other rules in other places (like using member variables within const methods or using `qgis::as_const` and then modifying the value of the original container).

**Comment by m-kuhn** [comment]
Thanks a lot @DelazJ

**Pull request merged**